packages/go_router/lib/src/route_data.dart

```diff
@@ -129,24 +148,132 @@ abstract class GoRouteData {
   );
 }
 
-/// Annotation for types that support typed routing.
+/// {@template shell_route_data}
+/// Baseclass for supporting
+/// [nested navigation](https://pub.dev/packages/go_router#nested-navigation)
+/// {@endtemplate}
+abstract class ShellRouteData extends RouteData {
+  /// {@macro shell_route_data}
+  const ShellRouteData();
+
+  /// [pageBuilder] is used to build the page
+  Page<void> pageBuilder(
+    BuildContext context,
+    GoRouterState state,
+    Widget navigator,
+  ) =>
+      const NoOpPage();
+
+  /// [pageBuilder] is used to build the page
+  Widget builder(
+    BuildContext context,
+    GoRouterState state,
+    Widget navigator,
+  ) =>
+      throw UnimplementedError(
+        'One of `builder` or `pageBuilder` must be implemented.',
+      );
+
+  /// A helper function used by generated code.
+  ///
+  /// Should not be used directly.
+  static ShellRoute $route<T extends ShellRouteData>({
+    required String path,
+    required T Function(GoRouterState) factory,
+    GlobalKey<NavigatorState>? navigatorKey,
+    List<RouteBase> routes = const <RouteBase>[],
+  }) {
+    T factoryImpl(GoRouterState state) {
+      final Object? extra = state.extra;
+
+      // If the "extra" value is of type `T` then we know it's the source
+      // instance of `GoRouteData`, so it doesn't need to be recreated.
+      if (extra is T) {
+        return extra;
+      }
+
+      return (_stateObjectExpando[state] ??= factory(state)) as T;
+    }
+
+    Widget builder(
+      BuildContext context,
+      GoRouterState state,
+      Widget navigator,
+    ) =>
+        factoryImpl(state).builder(
+          context,
+          state,
+          navigator,
+        );
+
+    Page<void> pageBuilder(
+      BuildContext context,
+      GoRouterState state,
+      Widget navigator,
+    ) =>
+        factoryImpl(state).pageBuilder(
+          context,
+          state,
+          navigator,
+        );
+
+    return ShellRoute(
+      builder: builder,
+      pageBuilder: pageBuilder,
+      routes: routes,
+      navigatorKey: navigatorKey,
+    );
+  }
+
+  /// Used to cache [ShellRouteData] that corresponds to a given [GoRouterState]
+  /// to minimize the number of times it has to be deserialized.
+  static final Expando<ShellRouteData> _stateObjectExpando =
+      Expando<ShellRouteData>(
+    'GoRouteState to ShellRouteData expando',
+  );
+}
+
+/// {@template typed_route}
+/// A superclass for each typed route descendant
+/// {@endtemplate}
+class TypedRoute<T extends RouteData> {
+  /// {@macro typed_route}
+  const TypedRoute({
+    this.routes = const <TypedRoute<RouteData>>[],
```

Comment:
Same here, the routes parameter should be duplicated in each subclasses instead because StatefulShellRoute is using branches instead of routes